xayn-ai/src/ranker/mod.rs

```diff
@@ -1,34 +1,55 @@
+mod context;
 pub(crate) mod public;
+mod utils;
 
+use std::collections::HashMap;
+
+use displaydoc::Display;
 use kpe::Pipeline as KPE;
+use thiserror::Error;
 
 use crate::{
-    coi::{point::UserInterests, CoiSystem},
+    coi::{compute_coi_for_embedding, point::UserInterests, Configuration},
+    data::document_data::CoiComponent,
     embedding::{smbert::SMBert, utils::Embedding},
     error::Error,
+    ranker::{
+        context::Context,
+        utils::{Document, Id as DocumentId},
+    },
+    utils::nan_safe_f32_cmp,
 };
 
+#[derive(Error, Debug, Display)]
+pub(crate) enum RankerError {
+    /// No user interests are known.
+    NoUserInterests,
+}
+
 /// The Ranker.
 pub(crate) struct Ranker {
-    #[allow(dead_code)]
-    coi: CoiSystem,
+    /// SMBert system.
     smbert: SMBert,
+    /// CoI configuration.
+    coi_config: Configuration,
```

Comment:
i still need the coi system for the feedback loop, but i agree that it makes sense to have the config directly in the ranker. maybe in another pr: we could move the `Configuration` from `coi::config::` to `ranker::config` and use it as the general config for the ranker, not only for the coi. or we could create a config for the ranker which holds a coi config next to possibly other configs. wdyt?

Reply:
I agree that is a good idea 👍